autocorpus/bioc_supplementary.py

```diff
@@ -0,0 +1,413 @@
+
+import datetime
+import re
+from pathlib import Path
+from typing import Any
+
+import pandas as pd
+from pandas import DataFrame
+
+from autocorpus.utils import replace_unicode
+
+from .bioc import (
+    BioCAnnotation,
+    BioCCollection,
+    BioCDocument,
+    BioCJSON,
+    BioCPassage,
+)
+from .bioc.bioctable import (
+    BioCTableCell,
+    BioCTableCollection,
+    BioCTableDocument,
+    BioCTableJSON,
+    BioCTablePassage,
+)
+
+
+def extract_table_from_pdf_text(text):
+    """Extracts tables from PDF text and returns the remaining text and tables.
+
+    Args:
+        text (str): The input text extracted from a PDF file.
+
+    Returns:
+        tuple: A tuple containing the remaining text (str) and a list of tables (list of DataFrames).
+    """
+    # Split the text into lines
+    lines = [x for x in text.splitlines() if x]
+    text_output = lines
+
+    # store extracted tables
+    tables = []
+    # Identify where the table starts and ends by looking for lines containing pipes
+    table_lines = []
+    # keep unmodified lines used in tables. These must be removed from the original text
+    lines_to_remove = []
+    inside_table = False
+    for line in lines:
+        if "|" in line:
+            inside_table = True
+            table_lines.append(line)
+            lines_to_remove.append(line)
+        elif (
+            inside_table
+        ):  # End of table if there's a blank line after lines with pipes
+            inside_table = False
+            tables.append(table_lines)
+            table_lines = []
+            continue
+
+    for line in lines_to_remove:
+        text_output.remove(line)
+
+    tables_output = []
+    # Remove lines that are just dashes (table separators)
+    for table in tables:
+        table = [line for line in table if not re.match(r"^\s*-+\s*$", line)]
+
+        # Extract rows from the identified table lines
+        rows = []
+        for line in table:
+            # Match only lines that look like table rows (contain pipes)
+            if re.search(r"\|", line):
+                # Split the line into cells using the pipe delimiter and strip whitespace
+                cells = [
+                    cell.strip()
+                    for cell in line.split("|")
+                    if not all(x in "|-" for x in cell)
+                ]
+                if cells:
+                    rows.append(cells)
+
+        # Determine the maximum number of columns in the table
+        num_columns = max(len(row) for row in rows)
+
+        # Pad rows with missing cells to ensure they all have the same length
+        for row in rows:
+            while len(row) < num_columns:
+                row.append("")
+
+        # Create a DataFrame from the rows
+        df = pd.DataFrame(rows[1:], columns=rows[0])
+        tables_output.append(df)
+    text_output = "\n\n".join(text_output)
+    return text_output, tables_output
+
+
+class BioCTableConverter:
```

Comment:
I think this and `BioCTextConverter` should both be dataclasses, like how I've suggested for the `bioc` package